File: runtimes/node/src/gateway.ts
class TenantLimiter {
    private tokens: number;
    private lastRefill: number;

    constructor(private readonly capacity: number, private readonly rate: number) {
        this.tokens = capacity;
        this.lastRefill = Date.now();
    }

    public consume(count: number): boolean {
        const now = Date.now();
        const elapsed = (now - this.lastRefill) / 1000.0;
        this.tokens = Math.min(this.capacity, this.tokens + elapsed * this.rate);
        this.lastRefill = now;

        if (this.tokens >= count) {
            this.tokens -= count;
            return true;
        }
        return false;
    }
}

export class MultiTenantRateLimiter {
    private readonly limiters: Map<string, TenantLimiter> = new Map();

    constructor(private readonly defaultCapacity: number, private readonly defaultRate: number) {}

    public setTenantLimit(tenantId: string, capacity: number, rate: number): void {
        this.limiters.set(tenantId, new TenantLimiter(capacity, rate));
    }

    public allow(tenantId: string, count: number): boolean {
        let limiter = this.limiters.get(tenantId);
        if (!limiter) {
            limiter = new TenantLimiter(this.defaultCapacity, this.defaultRate);
            this.limiters.set(tenantId, limiter);
        }
        return limiter.consume(count);
    }
}

File: runtimes/rust/src/gateway.rs
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::Instant;

pub struct TenantLimiter {
    capacity: f64,
    rate: f64,
    tokens: f64,
    last_refill: Instant,
}

impl TenantLimiter {
    pub fn new(capacity: f64, rate: f64) -> Self {
        Self {
            capacity,
            rate,
            tokens: capacity,
            last_refill: Instant::now(),
        }
    }

    pub fn consume(&mut self, count: f64) -> bool {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last_refill).as_secs_f64();
        self.tokens = (self.tokens + elapsed * self.rate).min(self.capacity);
        self.last_refill = now;

        if self.tokens >= count {
            self.tokens -= count;
            true
        } else {
            false
        }
    }
}

pub struct MultiTenantRateLimiter {
    limiters: Mutex<HashMap<String, TenantLimiter>>,
    default_capacity: f64,
    default_rate: f64,
}

impl MultiTenantRateLimiter {
    pub fn new(default_capacity: f64, default_rate: f64) -> Self {
        Self {
            limiters: Mutex::new(HashMap::new()),
            default_capacity,
            default_rate,
        }
    }

    pub fn set_tenant_limit(&self, tenant_id: &str, capacity: f64, rate: f64) {
        let mut guard = self.limiters.lock().unwrap();
        guard.insert(tenant_id.to_string(), TenantLimiter::new(capacity, rate));
    }

    pub fn allow(&self, tenant_id: &str, count: f64) -> bool {
        let mut guard = self.limiters.lock().unwrap();
        let limiter = guard.entry(tenant_id.to_string()).or_insert_with(|| {
            TenantLimiter::new(self.default_capacity, self.default_rate)
        });
        limiter.consume(count)
    }
}
